Can now specify idSeparator to control how id is created

green7ea opened the pull request:
By default, '_' is used as a separator for the different parts that
make up a field's id. This can be problematic when '_' is part of a
field's name.

This commit adds the ability to specify idSeparator when building a
form which allows users use a character that isn't used in field
names.

With the idSeparator option set to '.', we would get:

'root.field_a.field_123' instead of 'root_field_a_field_123'.

### Reasons for making this change

We use `_` as part of our field names and need to be able to access the complete path of nested fields in our internal application.

- relates to #531
- fixes #688

### Checklist

* [x] **I'm updating documentation**
  - [X] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [X] **I'm adding or updating code**
  - [X] I've added and/or updated tests
  - [x] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

green7ea commented:
This commit follows `idPrefix` closely, it didn't update the docs since `idPrefix` isn't mentioned there either.

epicfaace reviewed (commented):
Thanks, this is great! idPrefix is mentioned in the docs (see https://react-jsonschema-form.readthedocs.io/en/latest/api-reference/form-props/#idprefix), so can you please add a section for idSeparator as well?

green7ea commented:
Thanks for pointing out the documentation, I don't know how I missed it. I updated the commit:

- added documentation,
- added a line in the changelog for `v3.3.1 (upcoming)` pointing here.

I'm not sure if v3.3.1 was the right place to add the change so please let me know if I should move it.

heath-freenome commented:
This is awesome! I can use this fix as well

epicfaace commented on packages/core/test/utils_test.js:
```diff
@@ -2716,6 +2716,27 @@ describe("utils", () => {
+        bar: { $id: "rjsf/bar" },
+      });
+    });
+
```
Thanks! One more thing. Can you add a test within Form_test.js that sets idSeparator and makes sure that the fields are rendered with the right idSeparator?

heath-freenome replied:
@green7ea do you have time for adding this? Just one small change away from merging... if not, I can try

epicfaace reviewed (approved):
lgtm
